[P050] Additional output settings

tonhuisman opened the pull request:
- Added ~~minimal~~ updates to Adafruit library (including updated comments)
- Added Color Temperature DN40 calculated value
- Added Lux calculated value
- Added Clear output value
- Added Normalized RGB values (0..255)
- Added RGB Transformation matrix (3x3)
- Added optional events for calculated and measured values
- Cleaned up the existing options

TODO:
- [x] ~~Documentation~~ (done separately)
- [x] Adjust/fine-tune calculations (see linked issue)
- [x] Move status from DEVELOPMENT to TESTING

Resolves #3453 
Also resolves #426

tonhuisman committed:
[P050] Additional output settings
- Add Color Temperature DN40 calculation
- Add Lux calculation
- Add Clear output option
- Add Normalized RGB calculation

TD-er commented on src/_P050_TCS34725.ino:
```diff
@@ -162,17 +176,48 @@ boolean Plugin_050(byte function, struct EventStruct *event, String& string)
+        if (PCONFIG(2) == 1 && c != 0) { // R/G/B normalized to 0..255 (but avoid divide by 0)
+          UserVar[event->BaseVarIndex]     = (float)r / c * 255.0;
+          UserVar[event->BaseVarIndex + 1] = (float)g / c * 255.0;
+          UserVar[event->BaseVarIndex + 2] = (float)b / c * 255.0;
```
Please mark these as `255.0f` otherwise the compiler will add an extra `static_cast<float>(255.0)` equivalent to the code.

tonhuisman replied:
Ah, will fix, I copied that code almost verbatim from the Adafruit library.

TD-er commented on src/_P050_TCS34725.ino:
```diff
@@ -83,27 +84,40 @@ boolean Plugin_050(byte function, struct EventStruct *event, String& string)
+      optionsOutput[3] = F("Clear");
+      int optionValuesOutput[4] = { 0, 1, 2, 3};
+      addFormSelector(F("Output at Values #4"), F("p050_output4"), 4, optionsOutput, optionValuesOutput, PCONFIG(3));
+      addFormNote(F("Optionally adjust Values #4 name accordingly."));
```
Please place all these options + call to `addFormSelector` in their own scope (`{}`) so they can be destructed as soon as they are sent to the web client.

tonhuisman replied:
Yes, will fix. (but not before tonight, back at work from today 😃 )

TD-er committed:
Merge branch 'mega' into feature/P050_additional_output_settings

heinemannj reviewed (commented):
Please change
addFormCheckBox(F("Output calibrated RGB values"), F("p050_calibratedrgb"), PCONFIG(2) == 1, false);

to:
addFormCheckBox(F("Output normalized RGB values"), F("p050_calibratedrgb"), PCONFIG(2) == 1, false);

heinemannj reviewed (commented):
Compilation for esp8266 was successful.
Output values seems to be OK

tonhuisman committed:
[P050] Add Normalized sRGB calculation

tonhuisman committed:
Merge branch 'feature/P050_additional_output_settings' of https://github.com/tonhuisman/ESPEasy-1 into feature/P050_additional_output_settings

tonhuisman committed:
[P050] Apply suggested patches to the Adafruit library

heinemannj commented:
- Compilation for esp8266 was successful.
- Output values seems to be OK, but if these values are really valid - I have my strong doubts - to be discussed in the case
- GUI settings are really improved compared to your first version - for me the new look and feel is is good and understandable
- The documentation part should be my job - for sure ...

My proposal: Remove/rename the old and out of range Color Temperature function
The new DN40 calculation is really good and very closely to my LED datasheet:

Vendor datasheet             : 3000 K
Measured via DN40:         : 3061 K
Old function                     : 2140 K

Please crosscheck also the open pull request:
#https://github.com/letscontrolit/ESPEasy/pull/255
 The LED feature seems to be interesting
Otherwise please cancle the 255 request to avoid recurring reviews

Many thanks and take care!

TD-er commented:
Hmm a bit strange, but the sensor name/modelnr seems to be TCS3472 (as stated by the datasheet) but the naming by AdaFruit and in our plugin is TCS34725.
Maybe we should also rename it in the plugin?

Edit:
Hmm Digikey does seem to have the TCS34725, so maybe the [datasheet on Adafruit page](https://cdn-shop.adafruit.com/datasheets/TCS34725.pdf) is wrong?

Edit2:
the [latest datasheet on the site of the manufacturer](https://ams.com/documents/20143/36005/TCS3472_DS000390_3-00.pdf/6fe47e15-e32f-7fa7-03cb-22935da44b26) also mentions the nr without the '5'

heinemannj commented:
Additional comment/bug in Adafruit lib (cpp - you have implemented in plugin file):

Normally the clear value should be greater than the sum of (R+G+B) because of Infrared spectrum part which is not covered by RGB channels (especially under daylight conditions).
During my measurements with a Warm White LED in a dark room the sum of (R+G+B) was bigger then the Clear channel - auuuutsch ...

In general the usage of the clear channel make sense and is strongly proposed by the vendor ...
**BUT Clear must be NOT 0 and must be bigger then R+G+B** otherwise fallback to the sum of (R+G+B)

          case 1:
            if (c != 0) { // R/G/B normalized to 0..255 (but avoid divide by 0)
              UserVar[event->BaseVarIndex]     = (float)r / c * 255.0f;
              UserVar[event->BaseVarIndex + 1] = (float)g / c * 255.0f;
              UserVar[event->BaseVarIndex + 2] = (float)b / c * 255.0f;
            }
            break;
          case 2:
            if (c != 0) { // sR/G/B (normalized to 0.00..1.00 (but avoid divide by 0)
              UserVar[event->BaseVarIndex]     = (float)r / c * 1.0f;
              UserVar[event->BaseVarIndex + 1] = (float)g / c * 1.0f;
              UserVar[event->BaseVarIndex + 2] = (float)b / c * 1.0f;
            }

tonhuisman commented:
> the sensor name/modelnr seems to be TCS3472 (as stated by the datasheet) but the naming by AdaFruit and in our plugin is TCS34725.
> Maybe we should also rename it in the plugin?

Think I'll rename it, the official name seems to be TCS3472, the 5 looks like the manufacturer's name (AMS), where the last character is a partially rotated 5 that should probably be pronounced as an 'S' (see image below). As the chip itself isn't very exposed in all the offerings from Ali, and that's just a big copy/paste operation anyway, the TCS34725 name probably stuck.

Manufacturer logo:
![afbeelding](https://user-images.githubusercontent.com/1922393/104016121-ee11b380-51b5-11eb-9bc5-93e0dd348b86.png)

TD-er commented:
There also appears to be a TCS34727
So maybe those only differ slightly and the (programming) datasheet matches both?

tonhuisman commented:
> My proposal: Remove/rename the old and out of range Color Temperature function

For backward compatibility, I'd like to keep that, it's nearly no extra code.
I'll add `(deprecated)` to the selection combo for that setting.

> **BUT Clear must be NOT 0 and must be bigger then R+G+B** otherwise fallback to the sum of (R+G+B)

Didn't see that check yet, but it seems quite logical, so I'll add that.

tonhuisman commented:
> 
> 
> There also appears to be a TCS34727
> So maybe those only differ slightly and the (programming) datasheet matches both?

Hm, interesting, so maybe not a rename, or go for TCS3472x 😃 (and only in the device name, not the filename), unless it has f.e. a different physical or electrical interface.

TD-er commented:
> > There also appears to be a TCS34727
> > So maybe those only differ slightly and the (programming) datasheet matches both?
> 
> Hm, interesting, so maybe not a rename, or go for TCS3472x 😃 (and only in the device name, not the filename), unless it has f.e. a different physical or electrical interface.

Just searched through the datasheet for `TCS34725` and it seems there are only 2 differences:
- Reported ID differs (0x44 = TCS34725, 0x4D = TCS34727)
- Voltage levels for high/low on I2C bus are different. TCS34725 relates them to Vdd, where TCS34727 uses fixed voltages.

So keep the `TCS34725` name as the other one is less likely to be used and when changing it to `TCS3472x` a user may not find the chip as supported.

tonhuisman commented:
> Please crosscheck also the open pull request:
> ##255
> The LED feature seems to be interesting

That PR seems to mostly add interrupt handling, could be useful, but seems more in place for real-time processing, don't know if that would be applicable here.
For the LED support you could easily use a simple GPIO pin, don't like to duplicate that kind of code in any plugin, and I don't know if it would need some kind of acclimatization (delay) after turning on the LED.

heinemannj commented:
> For the LED support you could easily use a simple GPIO pin, don't like to duplicate that kind of code in any plugin, and I don't know if it would need some kind of acclimatization (delay) after turning on the LED.

Confirmed - I already use LEDs with serveral GPIO pins - then please remove the smelly old request ...

tonhuisman committed:
Merge branch 'mega' into feature/P050_additional_output_settings

tonhuisman committed:
[P050] Minor calculation adjustment, added to TESTING build

tonhuisman committed:
[P050] Adjusted Normalization calculation, added R/G/B Calibration factors

TD-er commented on src/_P050_TCS34725.ino:
```diff
@@ -106,16 +108,16 @@ boolean Plugin_050(byte function, struct EventStruct *event, String& string)
         optionsRGB[0] = F("Raw RGB");
         optionsRGB[1] = F("Normalized RGB (0..255)");
-        optionsRGB[2] = F("Normalized sRGB (0.00000..1.00000)");
+        optionsRGB[2] = F("Normalized rgb (0.00000..1.00000)");
```
Why is RGB here in lower case?
You also seem to differentiate between upper- and lower case in the note below.

tonhuisman replied:
That was a suggestion I took from @heinemannj in the linked issue #3453 but I'm not sure I like it that way, sRGB does look clearer to me. I'm not familiar with the custom abbreviations in this color-based field.

TD-er replied:
sRGB is a color space definition.
![image](https://user-images.githubusercontent.com/3751318/104107224-00294a00-52bb-11eb-83c8-aefc965757b8.png)
And the range (0....1) is an X/Y coordinate in the CIE1931 diagram. (N.B. it is 2-dimensional)

See also [this post on Eizo.nl](https://www.eizo.nl/praktische-informatie/knowhow-over-monitoren/kennis-over-kleurruimtes/) (not read it, but looks to go into details at first glance)

tonhuisman replied:
> looks to go into details at first glance

It does go deeper then I want to read into right now 😛.
Guess I'll be reverting rgb to sRGB then, np.

heinemannj replied:
- RGB is a color system / model
- sRGB is an RGB model defined by Microsoft
- rgb are normalized/scaled RGB values which are used for further transformation into other color systems

http://www.brucelindbloom.com/index.html?ColorCalculator.html

Please crosscheck: https://medium.com/hipster-color-science/a-beginners-guide-to-colorimetry-401f1830b65a

![image](https://user-images.githubusercontent.com/3251642/104114535-d5181800-5305-11eb-94a4-34c083ff591b.png)

![image](https://user-images.githubusercontent.com/3251642/104114477-2b388b80-5305-11eb-8f35-9723aa86181a.png)

heinemannj replied:
```
          case 2:
            if (t != 0) { // r/g/b (normalized to 0.00..1.00 (but avoid divide by 0)
              UserVar[event->BaseVarIndex]     = (float)r / t * 1.0f * calRed;
              UserVar[event->BaseVarIndex + 1] = (float)g / t * 1.0f * calGreen;
              UserVar[event->BaseVarIndex + 2] = (float)b / t * 1.0f * calBlue;
            }
            break;
```
You are right - If you apply calibration on RGB values you have to apply the same calibration on rgb values too.

But on the other hand rgb are normalized/scaled RGB values which are only used for further transformation into other color systems.

=>rgb values from normal user perspective are not really relevant and reviewing our discussion this kind of redundant data/information potentially open recurring discussions and questions.

If somebody needs to transform RGB values into other color spaces he can calculate rgb values from the calibrated RGB values by himself or can use a professional Color Calculator
http://www.brucelindbloom.com/index.html?ColorCalculator.html

The most important exercise was to fix the Adafriut bug for normalized RGB calculation - meaning not to divide through the Clear channel but to divide through to sum of RBG values that the Sum of rgb will remain 1.

My conclusion:
- We can securely remove the redandant rgb part without losing any precision and information

Sorry for the additional work and time but I'm also on a long journey to understand Colorimetry.

TD-er replied:
A lot of color spaces do take the color in 2 dimensions and the luminocity in a 3rd, precisely for this reason.
For example YUV and HSV are such color spaces.

And to give a very short introduction to the horse shoe shaped curve you see in the CIE1931 color space...
The outer curved line (not the straight line at the bottom) represent colors of a single wave length (typically reproducible using lasers)
This means the colored part of the chart is the range of colors we humans can see.

![image](https://user-images.githubusercontent.com/3751318/104121995-aec49d80-5342-11eb-9b5f-7f7a720ef9d1.png)

If you take 3 (or more) points on that chart (regardless whether they are inside or outside the chart), you can create all colors within that area of those points by mixing light (this chart is about light, not paint!!).
So to get back of the sRGB and Adobe color spaces (and there are more defined like NTSC, etc.) you see they are defined using coordinates in the XY plane of that chart.
Meaning to create a sRGB monitor, you should use primary colors of exactly those coordinates.
That's nearly impossible, so that's why monitor specs often refer to NN% of some color space when they mention the gamut (= the range of colors you can reproduce)

Typically "white" is around (1/3 , 1/3), but you also often see this chart with a curved line indicating a range of "daylight colors", like 5500K for example.
![image](https://user-images.githubusercontent.com/3751318/104122047-fea36480-5342-11eb-804c-a4a3a2dad307.png)

N.B. the blue numbers around the monochromatic curve are wavelength in nm.

To get an idea of what "infinite" color temperature looks like...
That's just about the color of the sky at a very cold sunny winter day.

TD-er commented:
By the way, we already have coded some functions in ESPEasy to convert between color spaces.

tonhuisman commented:
I found some RGBtoHSV and HSVtoRGB conversions, and calculations quite similar to this plugin in P066 VEML6040, but I'll leave it to an expert to suggest any more conversions to be added here, if any.
I've removed the 'rgb' selection for Red/Green/Blue output (the 0.0 to 1.0 range), as suggested by heinemannj as the only difference with the RGB setting is a factor 255, and changed 'Raw RGB' to 'Raw RGB (uncalibrated)' in the selection combo.

tonhuisman committed:
[P050] Removed unneeded output selection option

heinemannj commented:
I had some painful exercises with calibration via an general transformation matrix.

To make this story short and following best practices provided by AMS **a calibration matrix should be an 3x3 matrix**.
https://ams.com/documents/20143/80162/ColorSensors_AN000518_1-00.pdf

![image](https://user-images.githubusercontent.com/3251642/104287356-9d9c9d80-54b6-11eb-91e0-b63660970761.png)

This an example for an transformation matrix for adapting the tcs34725 RGB values to CIE XYZ "ColorChecker" references:

![image](https://user-images.githubusercontent.com/3251642/104287527-dfc5df00-54b6-11eb-8439-0e1b22875949.png)

With this approach I am able to patch a set of 24 reference colors from the "ColorChecker":

![image](https://user-images.githubusercontent.com/3251642/104285565-2108bf80-54b4-11eb-9fcb-521c2139b9e0.png)
![image](https://user-images.githubusercontent.com/3251642/104285629-3251cc00-54b4-11eb-959d-b9c364ba6e34.png)


This result is not perfect, but my measurement for sure was more than quick and dirty:

- ColorChecker image displayed on an DELL LCD monitor and measurement by putting the tcs34725 1cm above the display surface
- tcs34725 is inside a black box to avoid that any other light source will influence the measurement

TD-er commented:
What do you use as light source?

Edit:
never mind, a Dell monitor..
I thought you had one of those color charts.

Still, even on LCD, it does matter what the light source is.
For example older LCDs with those CCFL backlights tend to shift to purple when they age (due to loss of fosphor inside the tubes)
Also local dimming in modern LCDs may cause issues.

But I guess we could automate this in the web UI.

heinemannj commented:
I have already a working GUI (javascript) running on the ESPEasy webserver - a kind of ESPEasy Dashboard...
https://github.com/smartaquametering/smartPhotometer/blob/main/setup/configuration/colorimeter.html

A running demo for a photometer comparable to a colorimeter:
https://smartaquametering.github.io/photometer.html

Maybe a good approach in general to seggregate between tasks on backend and frontend.

TD-er commented:
Yep, that's also a good approach, but I wonder how well the response is when automating the calibration.
But it is for sure a good idea to have those things separated, to keep the build size reasonable in size.

heinemannj commented:
Example for measurement:
colored liquid solution in a cuvette lighted by an white D50 led - beam angel 180o

If you have a light yellow/brown colored solution and you wants to measure the Iodine Color Number
(mg of Jodine per 100ml potassium iodide solution (DIN 6162)
you have to make a proper calibration for  Iodine Color Numbers which will result in a dedicated transformation matrix

If you wants to refer to other Visual Color Scales additional dedicated transformation matrix needs to be setup.

The usage of the proper transformation matrix can be archived by selecting a predefined Visual Color Scale within the GUI

TD-er commented:
Hmm for measurements like this, you may also want to consider using a polarizer filter to remove reflection effects from the measurement as those may contribute quite a bit as a slight offset in intensity may cause quite a shift in recorded color.

TD-er committed:
Merge branch 'mega' into feature/P050_additional_output_settings

TD-er commented:
What is the status of this PR? mergable?

tonhuisman commented:
It's mergable, but could use documentation, that I'm hoping @heinemannj will add (but that would be a separate PR then)

TD-er commented:
OK, then let him start with the documentation.
No merge until it is documented, as I expect it will probably change a bit more when someone starts to document it and thus sees things may be done a bit more simple to explain .
Documentation is also a tool to see if something can be explained. If not, then it isn't good enough. :)

heinemannj commented:
> As mentioned earlier according to best practices provided by AMS **a calibration matrix should be an 3x3 matrix**:
> https://ams.com/documents/20143/80162/ColorSensors_AN000518_1-00.pdf
> 
> ![image](https://user-images.githubusercontent.com/3251642/104287356-9d9c9d80-54b6-11eb-91e0-b63660970761.png)

```
  R' = a1 * r + a2 * g + a3 * b;
  G' = b1 * r + b2 * g + b3 * b;
  B' = c1 * r + c2 * g + c3 * b;
```
Default values should be:

```
| 1	0	0 |
| 0	1	0 |
| 0	0	1 |

a1 = 1;
a2 = 0;
a3 = 0;

b1 = 0;
b2 = 1;
b3 = 0;

c1 = 0;
c2 = 0;
c3 = 1;
```

My supporting documentation will be based on the 3x3 Matrix approach.

TD-er commented:
> My supporting documentation will be based on the 3x3 Matrix approach.

Just keep in mind, you don't need to dive into too much detail, unless it is needed for explaining some of course.
But if you can link to somewhere for more detailed information, please do.
For example, if you figured out a specific matrix, sure document it. But if it is something straight out of the sensor's documentation, please refer to it.

Plugin documentation should be needed for using it, so try explain it from the view point of a potential user.
If some design decisions were made which may cause some artifacts in special use cases then it is of course needed to dive into it a bit more.

heinemannj commented:
> Plugin documentation should be needed for using it, so try explain it from the view point of a potential user.
> If some design decisions were made which may cause some artifacts in special use cases then it is of course needed to dive into it a bit more.

First documentation approach:
https://github.com/letscontrolit/ESPEasy/issues/3453#issuecomment-757332263

tonhuisman commented:
That's looking quite fine after your recent edit/cleanup, and now to pour that into RTD/Sphinx format 😉 (that's usually my biggest struggle)

tonhuisman committed:
Merge branch 'mega' into feature/P050_additional_output_settings

heinemannj commented:
> That's looking quite fine after your recent edit/cleanup, and now to pour that into RTD/Sphinx format 😉 (that's usually my biggest struggle)

Have you changed the calibration part into a 3x3 Matrix?

>  R' = a1 * r + a2 * g + a3 * b;
>  G' = b1 * r + b2 * g + b3 * b;
>  B' = c1 * r + c2 * g + c3 * b;

tonhuisman commented:
> Have you changed the calibration part into a 3x3 Matrix?

No, I haven't, was awaiting your guidance/request, so I'll take this as that request. Probably able to work on it tonight ⏳ 
May have some questions though.
Edit:
Assume the a1/2/3 etc. values are to be floats? Should there be checks on the value?

heinemannj commented:
> No, I haven't, was awaiting your guidance/request, so I'll take this as that request. Probably able to work on it tonight ⏳
> May have some questions though.
> Edit:
> Assume the a1/2/3 etc. values are to be floats?

a1: float, can also be =0
a2: float, can also be =0
a3: float, can also be =0

b1: float, can also be =0
b2: float, can also be =0
b3: float, can also be =0

c1: float, can also be =0
c2: float, can also be =0
c3: float, can also be =0

Please see the last example in my proposed documentation

>  Should there be checks on the value(s)?

ESPEasy based checkes on the values - dosn't make sense

Clear note for the user's:
Turn your head on before any modifications on the default calibration values ...

tonhuisman committed:
[P050] Added 3x3 matrix calibration, added opt. event for all values, adjusted device settings

tonhuisman commented:
- Added the 3x3 matrix calibration
- Added Reset calibration matrix option
- Added optional events for all Value 4 values
- Moved stuff to PluginStructs where applicable (and to store the calibration matrix)

Screenshot of initial device settings (only name entered, defaults now more sensible then all 0's):

![P050_InitialSettings](https://user-images.githubusercontent.com/1922393/104817078-4af91380-581f-11eb-94bb-c95822411ff4.png)

tonhuisman committed:
[P050] Fix missing include

tonhuisman committed:
Merge branch 'mega' into feature/P050_additional_output_settings

heinemannj commented:
**Transformation from RGB raw data into CIE XYZ (D65):**
```
| X |           | R |
| Y | = | M | * | G |
| Z |           | B |
```
**Calibration Matrix calculated by quick and dirty measurements using my not properly calibrated PC monitor and all of 24 patches of ColorChecker image:**

![image](https://user-images.githubusercontent.com/3251642/105010164-abfd3300-5a3b-11eb-9d61-8281ead40eb2.png)

**Modify ESPEasy settings:**

![image](https://user-images.githubusercontent.com/3251642/105010385-f088ce80-5a3b-11eb-96d0-a1e0848c16a4.png)

**Quick and dirty test with above transformation matrix:**

![image](https://user-images.githubusercontent.com/3251642/105008667-d0f0a680-5a39-11eb-8af9-8f8442212e7e.png)

**Crosscheck via CIE calculator:**

![image](https://user-images.githubusercontent.com/3251642/105010757-65f49f00-5a3c-11eb-88d1-eaec8f09b53d.png)

**Conclusion:**

The actual ESPEasy implementation is working for me in both cases:

- Using the raw data and performing all transformations on a frontend application
- Using ESPEasy transformation options

Based on my specific setup and calibration procedure I have an **CIE 1976 ΔE*ab average deviation of 3.6**
The **maximum CIE 1976 ΔE*ab is 6.3**.

Not so bad for a less then 10€ sensor ...

**Code issue in P050_data_struct.cpp:**

![image](https://user-images.githubusercontent.com/3251642/105012333-54ac9200-5a3e-11eb-9a7b-9d78a4b10aa6.png)

**Improvements:**

- Please change the wordings from 'Callibration' into 'Transformation'
- Output RGB Values: 'Raw RGB (uncalibrated)' => 'Raw RGB'
- The usage of Formulas in addtion to the transformation matrix is dangerous and make really NO sense

![image](https://user-images.githubusercontent.com/3251642/105013177-414df680-5a3f-11eb-88bd-0fa78d120252.png)

![image](https://user-images.githubusercontent.com/3251642/105016803-5f1d5a80-5a43-11eb-9744-4ce1bab9ff0b.png)

**Good job - may thanks!**

tonhuisman commented:
Great report!

> **Code issue in P050_data_struct.cpp:**
> 
> ![image](https://user-images.githubusercontent.com/3251642/105012333-54ac9200-5a3e-11eb-9a7b-9d78a4b10aa6.png)

Please elaborate, as I don't see what you are trying to point out, I think I'm multiplying `b3 * b` (= `matrix[1][2] * (float)b`) here

The changes and improvements I'll apply asap

heinemannj commented:
> Great report!
> 
> > **Code issue in P050_data_struct.cpp:**
> > ![image](https://user-images.githubusercontent.com/3251642/105012333-54ac9200-5a3e-11eb-9a7b-9d78a4b10aa6.png)
> 
> Please elaborate, as I don't see what you are trying to point out, I think I'm multiplying `b3 * b` (= `matrix[1][2] * (float)b`) here
> 
> The changes and improvements I'll apply asap

Please change with this corrected version (only a typo in the last matrix colomn:

```
/**
 * applyCalibration : calibrate r/g/b inputs (uint16_t) to rc/gc/bc outputs (float, by reference)
 */
void P050_data_struct::applyCalibration(uint16_t r, uint16_t g, uint16_t b, float *rc, float *gc, float *bc) {
  *rc = CalibrationSettings.matrix[0][0] * (float)r + CalibrationSettings.matrix[0][1] * (float)g + CalibrationSettings.matrix[0][2] * (float)b;
  *gc = CalibrationSettings.matrix[1][0] * (float)r + CalibrationSettings.matrix[1][1] * (float)g + CalibrationSettings.matrix[1][2] * (float)b;
  *bc = CalibrationSettings.matrix[2][0] * (float)r + CalibrationSettings.matrix[2][1] * (float)g + CalibrationSettings.matrix[2][2] * (float)b;
}
```

tonhuisman commented:
Ah, I didn't see _my_ typo as you showed only the corrected code, will fix that too of course

heinemannj commented:
I have identified that you have removed the normalized RGB option:

![image](https://user-images.githubusercontent.com/3251642/105031529-b75d5800-5a55-11eb-901e-a8d148633539.png)
2nd column in my excel sheet - calculated via Rraw / sum(Rraw, Graw, Braw) , ...

Even if this sRGB calculation is far away from the real sRGB value, this simple function provides a basic color estimation and is part of the origional Adafruit code

Maybe we can use the Formular again to make this calculation but I'm lost for getting it working

![image](https://user-images.githubusercontent.com/3251642/105033991-26887b80-5a59-11eb-8347-3e2023b8c493.png)

TD-er commented:
I notice there is not a lot of documentation on formula syntax.
For a simple example, see [here](https://espeasy.readthedocs.io/en/latest/Plugin/P004.html?highlight=formula#values)
A formula is normally used only on the value (or previous value) itself.
But I don't know why `[taskname#varname]` syntax could not be used there.
Just realize the output value is the result of such a formula, so the example you show cannot be used here as each value depends on the other ones.

If you need this, you should do this in the rules and assign the values to a variable.

TD-er commented:
I notice there is not a lot of documentation on formula syntax.
For a simple example, see [here](https://espeasy.readthedocs.io/en/latest/Plugin/P004.html?highlight=formula#values)
A formula is normally used only on the value (or previous value) itself.
But I don't know why `[taskname#varname]` syntax could not be used there.
Just realize the output value is the result of such a formula, so the example you show cannot be used here as each value depends on the other ones.

If you need this, you should do this in the rules and assign the values to a variable.

tonhuisman commented:
> you have removed the normalized RGB option

I got the impression the normalized values where no longer needed, but I can easily re-add them, just a small piece of code.

heinemannj commented:
> > you have removed the normalized RGB option
> 
> I got the impression the normalized values where no longer needed, but I can easily re-add them, just a small piece of code.

You are right - I will not use this normalized values any more!
I will only use transformated or raw data values in future.
It's up to you ...

The mentioned possibility to use rules and a dummy device is also working fine!

But then we should really remove the formula to avoid impact with transformation matrix

TD-er commented:
> But then we should really remove the formula to avoid impact with transformation matrix

I don't know if there are other use cases for this sensor?
What if someone else is using this sensor for something else than feeding it into a matrix for color conversion?

heinemannj commented:
Again - it's up to you to include this function again ...

**Conclusion:**
Based on my tests the **most trustable results** are related to (in below ordering):

**1. Transform raw data into CIE XYZ (Average ΔE*ab: 3.59 - Max: 6.31)**
2. Transform normalized RGB data into CIE XYZ (Average ΔE*ab: 17.23 - Max: 36.46)
3. Transform normalized RGB data into sRGB (Average ΔE*ab: 17.97 - Max: 42.42)
4. Transform raw data into sRGB (Average ΔE*ab: 18.00 - Max: 36.44)
5. No Transformation - use normalized RGB as sRGB (Average ΔE*ab: 23.46 - Max: 60.4)

**Normalized RGB as sRGB - raw estimation of colors ONLY** - because the sum of Normalized RGB at every time is: **255**

Example for Yellow - ColorChecker patch 16- sRGB(238 198 20):
```
Rraw: 3486.0
Graw: 2777.2
Braw: 1163.2
Sum : 7426.4

r: 3486.0 / 7426.4 = 0.469406
g: 2777.2 / 7426.4 = 0.373963
b: 1163.2 / 7426.4 = 0.156630
-----------------------------
                     1

R: 0.469406 * 255 = 120
G: 0.373963 * 255 = 95
B: 0.156630 * 255 = 40
-----------------------------
                    255
```
Above is even better then Adafruit source code - Adafruit is using as sum the Clear channel ...

**Sum of pure sRGB white is:**

```
R: 255
G: 255
B: 255
```
Autsssssch ...

This conclusion should be part of the ESPEasy documentation.

We should also consider to kick Adafruit into the right direction ...

tonhuisman committed:
[P050] Add extra transformation options, fixed a typo

tonhuisman commented:
I've tried to add all RGB options mentioned, if there are too many, they can easily be removed again.

I thought about adding events for the calculations not selected (each with their 3 values for r/g/b), but I'm not sure if that would be helpful, and it would add a lot of events if also 'Generate all as events' for Value 4 is enabled.

Settings are now changed into this:

![Screenshot - 19_01_2021 , 21_46_52](https://user-images.githubusercontent.com/1922393/105092237-4c397300-5aa1-11eb-8a23-2ae331e3f98c.png)

And the options for Output RGB values:

![Screenshot - 19_01_2021 , 21_50_06](https://user-images.githubusercontent.com/1922393/105092146-2ca24a80-5aa1-11eb-8c18-355d5cae4ff5.png)

heinemannj commented:
**I've tested all possible settings - Result is as expected - Code seems to be fine**

For the normalized RGB topic I've found a helpful web page: https://aishack.in/tutorials/normalized-rgb/

`I suggest to change the wording: calibration -> transformation`


![image](https://user-images.githubusercontent.com/3251642/105163463-3b761500-5b14-11eb-80b9-f83443f34dcf.png)


`For the Note: Check plugin documentation (i) on how to calibrate and how to calculate transformation matrix.`

`The wording in VEML6040 plugin is a little bit more precise by adding the SI unit:`


![image](https://user-images.githubusercontent.com/3251642/105169311-a6771a00-5b1b-11eb-89c4-ab4029badced.png)


```
- Color Temperature (...) [K]
- Ambient Light [Lux]
- Clear channel

```

TD-er commented:
Where does the data of the transformation/calibration matrix come from?

heinemannj commented:
> Where does the data of the transformation/calibration matrix come from?

You mean the nine matrix values?

You will need at least 3 references and 3 sensor measurements.
The procedure I have described here:
https://github.com/letscontrolit/ESPEasy/issues/3453#issuecomment-757332263

Not the final doc but a starting point

TD-er commented:
I was thinking, we can later also create some SVG image with the right color space chart (e.g. the CIE 1931) to visually show the measured values in the chart.
Is that useful?

heinemannj commented:
**This would be perfect!!!**

On the other hand I have already a working ESPEasy dashboard which is reading the ESP every 5 sec.
Pure Javascript with external out of the box libs for converting into different color spaces - Culori

What we already have:

- The most complete integration of the tcs34725 with nearly real color measurement!
- My Matrix transform into pure CIE XYZ coordinates with a reliable color delta - I never seen this in tons of tcs34725 examples all over the internet ...

Paired with all other great ESPEasy functionality:

- Combineing intensity of light measurement with TSL2591 and VEML6070 as a Photometer
- Using Atlas pH-Meter, EC-Meter and ORP-Meter controlled by ESPEasy
- Temperature measurements
- ...

TD-er commented:
Don't forget the rules for automating the measurements and/or acting on measured data.

heinemannj commented:
> Don't forget the rules for automating the measurements and/or acting on measured data.

Maybe a not the best solution - but it works:

```
On System#Boot do
  Let,1,0
  Let,3,0						// Loop counter reset
  Let,5,0						// Average value
  Let,6,0
  Let,7,0
  Let,8,0
  Let,10,16						// Sample Switch GPIO
  Let,11,12						// LED
  Pulse,[INT#11],1,1000
endon

on ResetReadings do
  TaskValueSet,%eventvalue1%,1,0
  TaskValueSet,%eventvalue1%,2,0
  TaskValueSet,%eventvalue1%,3,0
  TaskValueSet,%eventvalue1%,4,0
endon

on UpdateReadings do
  TaskValueSet,2,%eventvalue1%,%eventvalue2%		// Intensity
  if %eventvalue1%=4
    GPIO,[INT#10],0					// Sample finished
  endif
endon

on Sample#State=1 do
  GPIO,[INT#11],1					// LED on
  asyncevent,ResetReadings=2
  Let,3,1
  TaskRun,1
endon

on TCS34725#Clear do
  if [INT#3]>0
    Let,1,[INT#1]+1					// Loop counter
    if [INT#1]>1 and [INT#1]<6				// Skip value 1
      Let,5,[VAR#5]+[TCS34725#Red]
      Let,6,[VAR#6]+[TCS34725#Green]
      Let,7,[VAR#7]+[TCS34725#Blue]
      Let,8,[VAR#8]+[TCS34725#Clear]
    elseif [INT#1]=6
      Let,5,([VAR#5]+[TCS34725#Red])/5			// Average for value 2,3,4,5,6
      Let,6,([VAR#6]+[TCS34725#Green])/5
      Let,7,([VAR#7]+[TCS34725#Blue])/5
      Let,8,([VAR#8]+[TCS34725#Clear])/5
      asyncevent,UpdateReadings=1,[VAR#5]
      asyncevent,UpdateReadings=2,[VAR#6]
      asyncevent,UpdateReadings=3,[VAR#7]
      asyncevent,UpdateReadings=4,[VAR#8]
      GPIO,[INT#11],0					// LED off
      Let,3,0						// Exit
      Let,1,0						// Loop counter reset
      Let,5,0						// Average value
      Let,6,0
      Let,7,0
      Let,8,0
    endif
    TaskRun,1						// Loop
  endif
endon
```

tonhuisman committed:
Merge branch 'mega' into feature/P050_additional_output_settings

tonhuisman committed:
[P050] Rename Calibration to Transformation, added optional events for RGB values

heinemannj reviewed (approved):
I've compiled again and checked the new version:

- The new wording is good
- No further issue identified

Big thanks

tonhuisman commented:
>     * No further issue identified

Have you tried to turn on the Events options? I'm interested to see how much load it puts on an ESP8266 (haven't received my sensor yet, so still can't test myself)

tonhuisman committed:
Merge branch 'mega' into feature/P050_additional_output_settings

TD-er commented:
Status of this PR?

tonhuisman commented:
> Status of this PR?

I'm ready, unless someone (@heinemannj ?) is waiting for _me_ to add the documentation? IMHO we agreed that he's going to create a separate PR for that... Hope we're not waiting on each other 😛

heinemannj commented:
I will try to create a pull request for the documentation part - but to be honest the proper documentation of color measurement means is the most tricky job ...

My activity for getting the proper skill is a never ending story ...

tonhuisman commented:
> My activity for getting the proper skill is a never ending story ...

Don't worry, this year I will be in the IT 'business' for 36 years (professionally), and I'm still learning 😆
